text/0625-helper-managers.md

```diff
@@ -0,0 +1,504 @@
+   Combined with their ability to be hold a self contained lifecycle, this makes
+   them a powerful tool for composing declarative behavior.
+
+4. They can be _destroyed_. They tie in naturally to the destruction APIs that
+   we have recently added to Ember, and that allows their lifecycle to be
+   managed in a self contained way.
+
+However, this pattern has some issues today, mostly stemming from the fact that
+they execute _during_ render, which is not an ideal time for side-effecting.
+They can also be abused to modify app state, which can lead to difficult to
+follow code paths reminiscent of observers. These issues stem from a mismatch
+between two different goals, the goal of calculating a result or value, and the
+goal of triggering side-effects.
+
+The `isScheduledEffect` capability would schedule side-effecting helpers to execute
+_after_ render, and would _disable_ Ember's state mutations while they were
+running. This would ensure that side-effecting helpers run at the optimal time,
+and do not enable antipatterns and complicated codepaths.
+
+## Detailed design
+
+This RFC proposes adding the `setHelperManager` and `capabilities` APIs,
+imported from `@ember/helper`. Like `setComponentManager` and
+`setModifierManager`, `setHelperManager` receives a callback that is passed
+the owner, and returns an instance of the helper manager. When a helper
+definition is resolved by Ember, it will look up the manager recursively on the
+definition's prototype chain until it finds a helper manager. If it does not
+find one, it will throw an error.
+
+```ts
+type HelperDefinition = object;
+
+export declare function setHelperManager(
+  factory: (owner: Owner) => HelperManager,
+  definition: HelperDefinition
+): HelperDefinition;
+```
+
+And like the `capabilities` functions for component and modifier managers, the
+`capabilities` function for helper managers receives a version string as the
+first parameter and a options object as the second with optional flags as
+booleans. It produces an opaque `HelperCapabilities` object, which is assigned
+to the helper manager.
+
+```ts
+interface HelperCapabilitiesOptions {
+  hasValue?: boolean;
+  hasDestroyable?: boolean;
+  isScheduledEffect?: boolean;
+}
+
+type HelperCapabilities = Opaque;
+
+export declare function capabilities(
+  version: string,
+  options?: HelperCapabilitiesOptions
+): HelperCapabilities;
+```
+
+Helper managers themselves have the following interface:
+
+```ts
+interface HelperManager<HelperStateBucket> {
+  capabilities: HelperCapabilities;
+
+  createHelper(definition: HelperDefinition, args: TemplateArgs): HelperStateBucket;
+
+  getValue?(bucket: HelperStateBucket, args: TemplateArgs): unknown;
+
+  runEffect?(bucket: HelperStateBucket, args: TemplateArgs): void;
+
+  getDestroyable?(bucket: HelperStateBucket): object;
+}
+```
+
+Let's dig into these hooks one by one:
+
+### Hooks
+
+#### `createHelper`
```

Comment:
The issue with autotracking the creation hook is that it would break precedent, and we don't really have a clear meaning for it. We could call `getValue` or `runEffect` the next time that dirties, but it is already effectively "dirty" because we _will_ call those hooks anyways, and we don't want to teardown and recreate the helper because then we're taking control away from the user (they can no longer control their own tracking semantics).

Instead, users can use the `cache` API to wrap any code that they want to autotracking in this hook, and they can do anything they want with that information. They could return a cached value in the first `getValue`, they could destroy and recreate the instance of the helper, etc.

Really, `createHelper` just means "you exist in the template now", IMO. We shouldn't define any semantics for it beyond that.

Reply:
FWIW, I wasn't trying to convince you to change it (though I think we should 😉), but the fact that it is missing from being explicitly stated is my fundamental problem here.

Reply:
I added a section noting that the `createHelper` function is not tracked, and strategies for how users can track instance creation if they want to. We can keep discussing whether or not the hook should be tracked, I think the biggest question there is if it is, what would it do when it invalidates?

Reply:
> I think the biggest question there is if it is, what would it do when it invalidates?

Ya, this is **exactly** the question. 

@krisselden - I think you had thoughts here, what do you think?